sklearn/datasets/base.py

```diff
@@ -764,3 +764,108 @@ def _pkl_filepath(*args, **kwargs):
         basename += py3_suffix
     new_args = args[:-1] + (basename + ext,)
     return join(*new_args)
+
+
+class partialURLOpener(urllib.FancyURLopener):
+    """
+    Override HTTP Error 206 (partial file being sent)
+    """
+    def http_error_206(self, url, fp, errcode, errmsg, headers, data=None):
+        # Ignore the expected "error" code
+        pass
+
+
+def md5(path):
+    """
+    Calculate the md5 hash of the file at path.
+
+    Parameters
+    -----------
+    path: String
+        Path of file to calculate MD5 hash of.
+
+    Returns
+    -------
+    md5hash : String
+        MD5 hash of the file at the provided path.
+
+    """
+
+    md5hash = hashlib.md5()
+    md5hash.update(open(path, 'rb').read())
+    return md5hash.hexdigest()
+
+
+def validate_file_md5(expected_checksum, path):
+    """
+    Compare the MD5 checksum of a file at a path with
+    an expected MD5 checksum. If they do not match,
+    remove the file at path and throw a ValueError.
+
+    Parameters
+    -----------
+    expected_checksum: String
+        Expected MD5 checksum of file at path.
+
+    path: String
+        Path of file to compare MD5 hash of.
+
+    """
+
+    if expected_checksum != md5(path):
+        # remove the corrupted file
+        remove(path)
+        raise ValueError("{} has an MD5 hash differing "
+                         "from expected, file may be "
+                         "corrupted.".format(path))
+
+
+def fetch_and_verify_dataset(URL, path, checksum):
+    """
+    Fetch a dataset from a URL and check the MD5 checksum to ensure
+    fetch was completed and the correct file was downloaded
+
+    Parameters
+    -----------
+    URL: String
+        URL to fetch the download from.
+
+    path: String
+        Path to save the file to.
+
+    checksum: String
+        MD5 checksum to verify against the data
+
+    """
+
+    existing_size = 0
+    resume_url_downloader = partialURLOpener()
+    path_temp = path + ".tmp"
+    if exists(path_temp):
+        # since path_temp exists, resume download
+        temp_file = open(path_temp, "ab")
+        # get the amount of path_temp we've downloaded
+        existing_size = getsize(path_temp)
+        print("Resuming download from previous temp file, "
+              "already have {} bytes".format(existing_size))
+        # Download only the remainder of the file
+        resume_url_downloader.addheader("Range", "bytes="
```

Comment:
>IIRC, servers are not required to support this, which may mean you're sent the file from scratch.
This is true, but it's part of the original HTTP/1.1 spec, which means that (virtually all?) most webservers support it. Beyond that, i do agree that its a good idea to implement a check to see if the header is supported and act accordingly

> I wonder whether implementing resuming is a bit much.
It could be very well be, I figured why not use the temp file if we're writing one anyway.